Annotate System.Text.Encoding.CodePages for nullable reference types

stephentoub opened the pull request:
Contributes to https://github.com/dotnet/corefx/issues/40623
cc: @buyaa-n, @safern, @krwq, @tarekgh

stephentoub committed:
Annotate System.Text.Encoding.CodePages for nullable reference types

tarekgh commented on src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs:
```diff
@@ -1498,8 +1498,7 @@ private ISO2022Modes CheckEscapeSequenceKR(byte[] bytes, int escapeCount)
                         }
 
                         // Stick it in decoder
-                        if (decoder != null)
```
>     if (decoder != null) [](start = 19, length = 25)

I am not sure if removing this is correct. if the method GetCharsCP52936 is always called with valid decoder, then we should make it as not nullable and clean the code which check decoder==null. If GetCharsCP52936 is called with nullable decoder, we'll need to keep this check then and just move the next few lines under the condition.

stephentoub replied:
@tarekgh, just before this there's this code:
```C#
if (decoder == null || ...)
{
    ...
    break;
}
```
How could `decoder` be `null` here?

tarekgh replied:
you are right. I didn't notice the break. so we are fine then. 

---
In reply to: [337629043](https://github.com/dotnet/corefx/pull/41991#discussion_r337629043) [](ancestors = 337629043)

tarekgh commented on src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs:
```diff
@@ -1576,8 +1575,7 @@ private ISO2022Modes CheckEscapeSequenceKR(byte[] bytes, int escapeCount)
                             break;
                         }
 
-                        if (decoder != null)
```
>decoder [](start = 28, length = 7)

same

safern commented on src/System.Text.Encoding.CodePages/src/System/Text/DecoderNLS.cs:
```diff
@@ -37,13 +37,6 @@ internal DecoderNLS(EncodingNLS encoding)
             Reset();
         }
 
-        // This is used by our child deserializers
```
Isn't this needed?

stephentoub replied:
The type isn't serializable in .NET Core.

stephentoub committed:
Fix .NET Core 2.0 build

I'm not sure why we still build for this as a target, but there a bunch of nullability failures in the build because Debug.Assert isn't annotated in .NET Core 2.0's surface area, and thus places we're relying on Debug.Assert to suppress nullable warnings are failing.
